api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
+import com.fasterxml.jackson.dataformat.yaml.YAMLMapper;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.CustomResource;
+import io.strimzi.crdgenerator.annotations.Crd;
+import io.sundr.builder.annotations.Buildable;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A description of a Kafka assembly, as exposed by the Strimzi Kafka CRD.
+ */
+@JsonDeserialize(
+        using = JsonDeserializer.None.class
+)
+@Crd(
+        apiVersion = KafkaAssembly.CRD_API_VERSION,
+        spec = @Crd.Spec(
+                names = @Crd.Spec.Names(
+                        kind = KafkaAssembly.RESOURCE_KIND,
+                        plural = KafkaAssembly.RESOURCE_PLURAL
+                ),
+                group = KafkaAssembly.RESOURCE_GROUP,
+                scope = "Namespaced",
+                version = KafkaAssembly.VERSION
+        )
+)
+@JsonPropertyOrder({"apiVersion", "kind", "metadata", "spec"})
+@Buildable(
+        editableEnabled = false,
+        generateBuilderPackage = true,
+        builderPackage = "io.strimzi.api.kafka.model"
+)
+public class KafkaAssembly extends CustomResource {
+
+
+    private static final long serialVersionUID = 1L;
+    public static final String VERSION = "v1alpha1";
+    public static final String RESOURCE_KIND = "Kafka";
+    public static final String RESOURCE_LIST_KIND = RESOURCE_KIND + "List";
+    public static final String RESOURCE_GROUP = "kafka.strimzi.io";
+    public static final String RESOURCE_PLURAL = "kafkas";
+    public static final String RESOURCE_SINGULAR = "kafka";
+    public static final String RESOURCE_NAME = RESOURCE_PLURAL + "." + RESOURCE_GROUP;
+    public static final String CRD_API_VERSION = "apiextensions.k8s.io/v1beta1";
+    public static final String CRD_NAME = RESOURCE_PLURAL + "." + RESOURCE_GROUP;
+
+    private String apiVersion;
+    private ObjectMeta metadata;
+    private transient KafkaAssemblySpec spec;
+    private Map<String, Object> additionalProperties = new HashMap<>(0);
```

Comment:
What's the purpose of this `additionalProperties` ?

Reply:
K8s takes the stance, and enforces it for CRDs, that extra properties added to the JSON (not defined in the schema) are ignored. You can configure Jackson to do the same, but then when Jackson deserializes a POJO tree it will discard the extra properties, which will then be missing when the POJO tree gets serialized. the additionalProperties (specifically the `@JsonAnyGetter` and `@JsonAnySetter` on the getter and setter) give Jackson somerwhere to put the properties which are not otherwise mapped and means we can round trip the properties we "don't know about" back to JSON.

I don't think this is 100% necessary right now, because we currently only ever deserialize the JSON. I think it will be a problem when:

* we come to support `status` in the CRDs (for which we might need the CRD subresource support in k8s/os, but I've not thought it through in depth yet).
* we want to evolve the CRD schema without incrementing the apiVersion, which is why K8s takes the stance it does.

Basically I think it makes sense if the POJOs support hi-fidelity marshalling to/from JSON right from the start.